perception/autoware_tensorrt_bevdet/src/nvjpegdecoder.cpp

```diff
@@ -0,0 +1,299 @@
+
+  CHECK_NVJPEG(nvjpegStateAttachPinnedBuffer(
+    params.nvjpeg_decoupled_state, params.pinned_buffers[buffer_index]));
+
+  CHECK_NVJPEG(nvjpegDecodeJpegHost(
+    share_param.nvjpeg_handle, params.nvjpeg_decoder, params.nvjpeg_decoupled_state,
+    params.nvjpeg_decode_params, params.jpeg_streams[buffer_index]));
+
+  CHECK_CUDA(cudaStreamSynchronize(params.stream));
+
+  CHECK_NVJPEG(nvjpegDecodeJpegTransferToDevice(
+    share_param.nvjpeg_handle, params.nvjpeg_decoder, params.nvjpeg_decoupled_state,
+    params.jpeg_streams[buffer_index], params.stream));
+
+  buffer_index = 1 - buffer_index;  // switch pinned buffer in pipeline mode to avoid an extra sync
+  RCLCPP_DEBUG(rclcpp::get_logger("nvjpegDecoder"), buffer_index);
+
+  CHECK_NVJPEG(nvjpegDecodeJpegDevice(
+    share_param.nvjpeg_handle, params.nvjpeg_decoder, params.nvjpeg_decoupled_state, &out,
+    params.stream));
+
+  // CHECK_CUDA(cudaStreamSynchronize(params.stream)); //TODO new add
+
+  CHECK_CUDA(cudaEventRecord(stopEvent, params.stream));
+
+  CHECK_CUDA(cudaEventSynchronize(stopEvent));
+  CHECK_CUDA(cudaEventElapsedTime(&loopTime, startEvent, stopEvent));
+  time = 0.001 * static_cast<double>(loopTime);  // cudaEventElapsedTime returns milliseconds
+                                                 // time = 1.0;
+}
+
+nvjpegDecoder::nvjpegDecoder(size_t n, size_t _fmt)
+: N_img(n), iout(n), isz(n), widths(n), heights(n), params(n), share_param(_fmt)
+{
+  init();
+}
+
+int nvjpegDecoder::decode(const std::vector<std::vector<char>> & files_data, uchar * out_imgs)
+{
+  for (size_t i = 0; i < params.size(); i++) {
+    CHECK_CUDA(cudaStreamCreate(&params[i].stream));
+  }
+  if (prepare_buffers(files_data, widths, heights, iout, isz, share_param)) return EXIT_FAILURE;
+
+  double times[6];
+
+  auto decode_start = high_resolution_clock::now();
+  std::vector<std::thread> threads(files_data.size());
+  for (size_t i = 0; i < files_data.size(); i++) {
+    threads[i] = std::thread(
+      decode_single_image, std::ref(files_data[i]), std::ref(iout[i]), std::ref(params[i]),
+      std::ref(share_param), std::ref(times[i]));
+  }
+  for (size_t i = 0; i < files_data.size(); i++) {
+    threads[i].join();
+  }
+  auto decode_end = high_resolution_clock::now();
+  duration<double> decode_time = decode_end - decode_start;
+
+  RCLCPP_INFO(
+    rclcpp::get_logger("nvjpegDecoder"), "Decode total time : %.4lf ms",
+    decode_time.count() * 1000);
+
+  for (size_t i = 0; i < files_data.size(); i++) {
+    get_img(
+      iout[i].channel[0], iout[i].pitch[0], iout[i].channel[1], iout[i].pitch[1],
+      iout[i].channel[2], iout[i].pitch[2], widths[i], heights[i],
+      out_imgs + i * 3 * widths[i] * heights[i]);
+  }
+
+  for (size_t i = 0; i < params.size(); i++) {
+    CHECK_CUDA(cudaStreamDestroy(params[i].stream));
+  }
+  return EXIT_SUCCESS;
+}
+
+int nvjpegDecoder::init()
+{
+  nvjpegDevAllocator_t dev_allocator = {&dev_malloc, &dev_free};
```

Comment:
While using custom allocators is fine for CUDA contexts,  ensure that these allocators (`dev_malloc`, `dev_free`, `host_malloc`, `host_free`) handle errors robustly.  Currently, only the return value is checked; consider adding error handling and logging for more informative error messages in the event of allocation failures.